### PR TITLE
net, rpc:  return NET_UNROUTABLE as not_publicly_routable, automate helps

### PR DESCRIPTION
**Pull request by jonatack**
per the IRC discussion today at http://www.erisian.com.au/bitcoin-core-dev/log-2021-01-19.html#l-87

- return a more helpful string name for `Network::NET_UNROUTABLE`: "not_publicly_routable" instead of "unroutable"
- update the RPC getpeerinfo "network" help, and automate it and the getnetworkinfo "network#name" and the -onlynet help doc generation

### CODE REVIEW

**Review by theStack — APPROVED** [review]
ACK d9ca41964dae2bc58ed8f9639847951beb0f3453

File: src/rpc/net.cpp [hunk 1]
```diff
@@ -587,7 +587,7 @@ static RPCHelpMan getnetworkinfo()
                         {
                             {RPCResult::Type::OBJ, "", "",
                             {
-                                {RPCResult::Type::STR, "name", "network (ipv4, ipv6 or onion)"},
+                                {RPCResult::Type::STR, "name", "network (ipv4, ipv6, onion, i2p, cjdns, or not_publicly_routable)"},
```
Comment (MarcoFalke) [inline src/rpc/net.cpp:590]: unrelated nit: Could return a list of all networks in netbase and call `Join(...)` here? Would simplify future review if a network is added or the list added to another place.

Comment (jonatack) [reply src/rpc/net.cpp]: proposed an update, not sure if it's too dynamic

Comment (jonatack) [reply src/rpc/net.cpp]: done, it works but not sure if it's kosher to iterate over the enum instead of using a static, manually updated array

Comment (jonatack) [reply src/rpc/net.cpp]: added regression coverage for the generated help

Comment (jonatack) [reply src/rpc/net.cpp]: (it seems ok to iterate over the enum per the doc in `src/netaddress.h`)
```
 * Keep these sequential starting from 0 and `NET_MAX` as the last entry.
 * We have loops like `for (int i = 0; i < NET_MAX; i++)` that expect to iterate
 * over all enum values and also `GetExtNetwork()` "extends" this enum by
 * introducing standalone constants starting from `NET_MAX`.
 */
enum Network
```

File: src/netbase.cpp [hunk 2]
```diff
@@ -68,6 +68,16 @@ std::string GetNetworkName(enum Network net)
     assert(false);
 }
 
+std::vector<std::string> GetNetworkNames()
+{
+    std::vector<std::string> names;
+    for (int n = Network::NET_IPV4; n != Network::NET_INTERNAL; ++n) {
+        names.emplace_back(GetNetworkName(static_cast<Network>(n)));
+    }
```
Comment (theStack) [inline src/netbase.cpp:76]: nit: I'd slightly prefer a more generic approach here for the loop that doesn't make any assumptions about the position of concrete enums:

```suggestion
    for (int n = 0; n < Network::NET_MAX; ++n) {
        enum Network network = static_cast<enum Network>(n);
        if (network == Network::NET_UNROUTABLE || network == Network::NET_INTERNAL) continue;
        names.emplace_back(GetNetworkName(network));
    }
```
(untested)

Comment (jonatack) [reply src/netbase.cpp:76]: Yes, that's better. Thanks!

Comment (jonatack) [reply src/netbase.cpp]: Done, and moved it next to GetNetworksInfo() as the 2 functions are similar.

**Comment by jonatack** [comment]
Updated per review feedback by @MarcoFalke and @theStack and added an optional commit to use named casts (can be dropped if preferred).

**Review by theStack — APPROVED** [review]
light re-ACK 4c97ff24eef77bde3c27258f93e8d216caac9cc8 🦉 

Reviewed the second commit via `--color-moved=dimmed-zebra` to verify that the changes involving `GetNetworksInfo` were move-only (apart from taking advantage of C++11 initializer list at one place) .
Changes LGTM, also replacing C-style casts with named casts is a good idea to enable compile time checks. My reason for only "light"-ACKing: I'm not 100% sure how dynamic the generation of RPC help texts is allowed to be, other uses of `Join(...)` in RPC help texts seem only to use constants:
```
$ git grep RPC.*Join
src/rpc/net.cpp:                            {RPCResult::Type::STR, "network", "Network (" + Join(GetNetworkNames(), ", ") + ")"},
src/rpc/net.cpp:                                {RPCResult::Type::STR, "permission_type", Join(NET_PERMISSIONS_DOC, ",\n") + ".\n"},
src/rpc/net.cpp:                            {RPCResult::Type::STR, "connection_type", "Type of connection: \n" + Join(CONNECTION_TYPE_DOC, ",\n") + ".\n"
src/rpc/net.cpp:                                {RPCResult::Type::STR, "name", "network (" + Join(GetNetworkNames(), ", ") + ")"},
```

**Comment by jonatack** [comment]
@theStack thanks--I asked myself the same question and decided to start with proposing dynamic generation because returning the help doc isn't a perf hotspot and, unlike the two existing cases, this one can be generated automatically (the functional tests enforce a sanity check on any upstream updates).

**Comment by MarcoFalke** [comment]
> replacing C-style casts with named casts is a good idea to enable compile time checks

What is an example for a compile time check that is enabled when using those for numeric types? I am only aware of compile time checks that are enabled for pointers. Static casts for numeric values are just overly verbose for no reason and I'd prefer not to change existing code.

**Comment by theStack** [comment]
> What is an example for a compile time check that is enabled when using those for numeric types?

If the destination type is numeric, the named cast asserts at compile time that also the source type is numeric, while a c-style cast doesn't in all cases:
```
long foo = (long)some_pointer; // compiles happily
long bar = static_cast<long>some_pointer; // "error: invalid ‘static_cast’ from type ‘char*’ to type ‘long int’"
```

Also note that modern C++ coding guidelines discourage the use of C casts:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-casts-named
https://google.github.io/styleguide/cppguide.html#Casting

That said, I'm happy to re-ACK this PR also without the third commit.

File: src/rpc/net.cpp [hunk 3]
```diff
@@ -237,13 +237,13 @@ static RPCHelpMan getpeerinfo()
...
-            obj.pushKV("pingtime", ((double)stats.m_ping_usec) / 1e6);
+            obj.pushKV("pingtime", static_cast<double>(stats.m_ping_usec) / 1e6);
         }
         if (stats.m_min_ping_usec < std::numeric_limits<int64_t>::max()) {
-            obj.pushKV("minping", ((double)stats.m_min_ping_usec) / 1e6);
+            obj.pushKV("minping", static_cast<double>(stats.m_min_ping_usec) / 1e6);
         }
         if (stats.m_ping_wait_usec > 0) {
-            obj.pushKV("pingwait", ((double)stats.m_ping_wait_usec) / 1e6);
+            obj.pushKV("pingwait", static_cast<double>(stats.m_ping_wait_usec) / 1e6);
```
Comment (MarcoFalke) [inline src/rpc/net.cpp:246]: Still not convinced that it is worth it to change existing code that is being removed in #21015

**Review by maflcko — COMMENTED** [review]
ok, fair enough on the static_cast, but changing existing code (especially one that is about to get removed) might not be worth it.

**Comment by jonatack** [comment]
Yup, dropped the last commit, it's not related to this change and can be done later when the relevant lines are touched. Thanks for reviewing!

**Review by theStack — APPROVED** [review]
re-ACK 23299def6acda0c07d14274c605dbd4c9d23fe56 🍏

File: src/rpc/net.cpp (hunk 3 above)
Comment (MarcoFalke) [reply src/rpc/net.cpp]: Generally, it might be better to use C++11 non-narrowing casts, or ideally no cast at all.

File: src/rpc/net.cpp [hunk 4]
```diff
@@ -587,7 +601,7 @@ static RPCHelpMan getnetworkinfo()
                         {
                             {RPCResult::Type::OBJ, "", "",
                             {
-                                {RPCResult::Type::STR, "name", "network (ipv4, ipv6 or onion)"},
+                                {RPCResult::Type::STR, "name", "network (" + Join(GetNetworkNames(), ", ") + ")"},
```
Comment (MarcoFalke) [inline src/rpc/net.cpp:604]: I don't think getnetworkinfo returns an object for unroutable networks? Also, i2p, etc aren't implemented yet.

File: src/rpc/net.cpp [hunk 5]
```diff
@@ -542,25 +575,6 @@ static RPCHelpMan getnettotals()
     };
 }
 
-static UniValue GetNetworksInfo()
-{
-    UniValue networks(UniValue::VARR);
-    for (int n = 0; n < NET_MAX; ++n) {
-        enum Network network = static_cast<enum Network>(n);
```
Comment (MarcoFalke) [inline src/rpc/net.cpp:549]: Any reason to add `enum` in this move-only? The compiler, as well as the dev already know this is an enum

**Review by maflcko — COMMENTED** [review]
left some feedback  on 23299def6acda0c07d14274c605dbd4c9d23fe56 🔌

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

left some feedback  on 23299def6acda0c07d14274c605dbd4c9d23fe56 🔌
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUipnAv/YQrIj8eF8B2E2U0kCmbK1ftlsNowtsHgXUIHRB+1Ow8i3c7ZsjYnQp7Z
CfdGVWgiQ/4UkWIXlimM+8IBkQnkggupIQN4w4/3Gm7i9R3HlC4AMQV+Qov7Jhpn
qPRf2jbt9sUJCYt0tW7sH/TvwZTjJrOkM9wy6PseK8u06bTTWdVWRGjl/bZ7cCt5
RZ0sJ01AQ4y7dDCMgbKPx/t120RdzPwFT/blSLXi0/SIWZNtfX7oXQ+nlKEJ5bIa
jeg/z5P88QkfWsbMbSERycZW3WbFA13o7Um39P/tSdzBzOa8ykXA/fHdJo7g8oC/
SOQk8oqSJuDsL7jEkM78JrBt4AIiIOSuIXnafb0nvZ1ZA9tDoRrxsK7bNV2HZ9nr
fKuy++V7LJsMwRzn6jLeZARkHGDT8RNpTnpDYOiCC5dGfCgxJhwQ/hZ0mcVDrTZd
nGRo63Fe7O3zKnnenNZRHB5m+yfuq0rslIveDgNCeRdmusGrc+UnGAYcxQO++z5J
8PopnI7e
=tod6
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `27f463ade22a48024739db23d1039da56bed75cd5631f55b35e5df1c8ebcbea7  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010827f463ade22a48024739db23d1039da56bed75cd5631f55b35e5df1c8ebcbea7f010ecc64edf5466464883ca0ec6e0d4e84708fff0105ba3f5856c4d9db22a6fbd51304eb6ff08f1046017cc28f0081dbf47df9543873f0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010710825e7570492293f5c0a186011129e08f1046017cc28f0089ee78d5c403ee2cb0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01081e0650a570ee7e929067e7d49198bd808f1046017cc28f008599cb5687bc446260083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010c47b077a195cbd4831710ad112d861b608f1046017be58f0085ca376f657ddc3040083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

File: src/rpc/net.cpp (hunk 5 above)
Comment (MarcoFalke) [reply src/rpc/net.cpp:549]: Oh, nvm you added `const`, not `enum`.

File: src/rpc/net.cpp (hunk 4 above)
Comment (jonatack) [reply src/rpc/net.cpp:606]: Good catch about getnetworkinfo! (and I should have remembered that, too). Fixed both.

**Comment by jonatack** [comment]
Updated per `git diff 23299de bbb07d1` for @MarcoFalke's feedback (thanks!)

**Comment by jonatack** [comment]
Realized that `-onlynet` in `src/init.cpp` should use `GetNetworkNames()` too. Rebased, as `-onlynet` was just changed, and updated.

**Review by theStack — APPROVED** [review]
re-ACK 96635e61777add29b6a34d47767a63f43b2919af 🌳 
Note that the exact string listing in parantheses changes slightly now, as the "or" part is not there anymore, but I think we can perfectly live with that.

**Comment by jonatack** [comment]
> Note that the exact string listing in parantheses changes slightly now, as the "or" part is not there anymore, but I think we can perfectly live with that.

Indeed. I hesitated to do something like
```cpp
names.emplace_back("or " + GetNetworkName(NET_UNROUTABLE));
```

**Review by maflcko — COMMENTED** [review]
review ACK 96635e61777add29b6a34d47767a63f43b2919af 🐗

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 96635e61777add29b6a34d47767a63f43b2919af 🐗
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUi1nAv8Co9dfz7Kuxi/uUXJ3ob+y0mhUj0d023QFAyW/2Wh/C+QKZu6HRs8Q8Qh
PG+LNvgT+xfvEZb3bYPUUalSmODqU4OkgkulK1YPCr4zxquExcSFkSFkJL8gZzRu
CyduWAzAfGf2I0lGopqgsHMpfs6Q7SYeoEyXRHjztEzZu2FWZMQEuU5APep5e4q9
tw9Qkw1cfScc0rhROs+GbRj6N8AxA0B+GvvKWvc7Y16RfCIaWlP4vKzwN0s18Lzb
JPKRTTKmasfGHkIgnPCKmR2nlu5z8PekfGtvo5wMrRC5fVsFxedgTrtVuaRBk/Jc
IOM7qdVHvp/ClMvbu78Dm9kTwPTs2yFvg9/XcN57vTNQb4Qme9DJiMzOtELvwgaE
xM3xbwskCTldH6CPwMU7QJtmJxHnsNuO0VmtD7BX4QdCpyGXVXwjoYVpN6yn+/BJ
Ad9EKVDN/OCiC+jXiidgQvy6aiUoev1CEH23zvk3RRLXp+HbqNQfLzJfk6deQ5uF
GAJxcJlJ
=v5MP
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `946105e795bca254ed011a4ffce2cf299168c362a6c99aed2702782f5738dc7f  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108946105e795bca254ed011a4ffce2cf299168c362a6c99aed2702782f5738dc7ff010a812a30129d7eedeed5476d2e318d3c608fff01009aa7422b09152035889f7a0ef6f5af808f104602a853df00805d72e162c121c710083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0101c10770d367e7b0f4356a39f7512ec0f08f104602a776bf008d20cbd436d1278370083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff01081d3eb13f57918227502bdb71de8257a08f104602a853df008bbd0ea70091dfbc30083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010a780f0279576d003dfc09bfecb28df5508f104602a853df008fdda897f2929308f0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

File: src/netbase.cpp [hunk 6]
```diff
@@ -68,6 +68,20 @@ std::string GetNetworkName(enum Network net)
     assert(false);
 }
 
+std::vector<std::string> GetNetworkNames(bool append_unroutable)
+{
+    std::vector<std::string> names;
+    for (int n = 0; n < NET_MAX; ++n) {
+        const enum Network network{static_cast<Network>(n)};
+        if (network == NET_UNROUTABLE || network == NET_I2P || network == NET_CJDNS || network == NET_INTERNAL) continue;
```
Comment (jonatack) [inline src/netbase.cpp:76]: @vasild you may want to update this line (and the feature_proxy tests) in #20685 if you're inclined to it in that pull

Comment (jonatack) [inline src/netbase.cpp:76]: @vasild you may want to update this line to remove `NET_I2P` (and add it to the rpc_net.py tests) in #20685 (if you're inclined to do it there)

Comment (vasild) [reply src/netbase.cpp:76]: Done in 433d9a9f1e416c5518f832b4ca9402c8807877bc `net: Do not skip the I2P network from GetNetworkNames()`, thanks for the notice!

**Comment by jonatack** [comment]
@shesek done, per your original IRC issue getpeerinfo's network field returns (and documents in the help) the `Network::NET_UNROUTABLE` type as "not_publicly_routable" (master/v22.0)

**Review by trongham — COMMENTED** [review]
Tr

**Pull request merged**